### ISSUE TITLE
[API Proposal]: Provide an `IUnknown` IID constant

### ISSUE THREAD

**Issue by AaronRobinsonMSFT** [dotnet/runtime#91150]
### Background and motivation

The `IUnknown` interface ID (IID) is a well known constant that has been set since the early 90's. This IID can be used in a myriad cases, particularly with the new COM source generator. This constant should be defined in a way that can be used both as a declaration on an interface definition and as a way to easily implement `QueryInterface()`.

Additional other IIDs that should be considered:

- `IID_IClassFactory`
- `IID_IClassFactory2`

### API Proposal

```diff
namespace System.Runtime.InteropServices;

public abstract class ComWrappers
{
+    public const Guid IID_IUnknown;
}
```

### API Usage

```csharp
// Fancy the value
var c = new MyFancyCollection<int>();
c.Fancy(42);

// Getting the values out
foreach (var v in c)
    Console.WriteLine(v);
```

### Alternative Designs

Define an `IUnknown` interface type in C# and define the value using the `GuidAttribute`. 

### Risks

None. This is a well known value and one that should be available.

**Comment by AaronRobinsonMSFT** [comment]
@tannergooding Your thoughts would be much appreciated on this proposal.

**Comment by Sergio0694** [comment]
Would it be possible to expose these as a `static ref readonly Guid` property like in TerraFX? That'd make them more flexible and marginally improve perf (no static ctor, just reading straight from RVA fields). I'm aware that the approach TerraFX uses is technically Windows specific due to alignment concerns, but given this is for .NET 9 could that be something we could fix, so that it can be used reliably and cross-platform? 🤔

**Comment by AaronRobinsonMSFT** [comment]
> marginally improve perf (no static ctor, just reading straight from RVA fields)

Do we have measurements that show this has a performance impact?

**Comment by Sergio0694** [comment]
I think Tanner has some, as he specifically had to remove them from TerraFX as they were piling up and causing a significant slowdown at startup (admittedly, he had thousands of those). Though to clarify, it's not just about performance, but also convenience, as having them return a `ref readonly` also means you can eg. pin and pass them by pointer, or pass them to a method taking `in`/`ref readonly` without needing to use a local temp.

**Comment by jkotas** [comment]
> The IUnknown interface ID (IID) is a well known constant that has been set since the early 90's

The same can be said about number of other COM related constants that one needs for most real-world COM scenarios. (For example, the `E_...` error codes.)

Our principle so far was to expose the absolute minimum APIs required for specific higher level interop variants. This would be a start of exposing additional surface for convenience. Where would we stop?

**Comment by jkotas** [comment]
> Alternative Designs
> Define an IUnknown interface type in C# and define the value using the GuidAttribute.

The COM interop packages out there use several different patterns:
- `ref readonly Guid` returning property in `static class IID`: https://github.com/terrafx/terrafx.interop.windows/blob/fadce5a41fa5e6f0282e80e96f033d0a2c130991/sources/Interop/Windows/Windows/um/Unknwnbase/IID.cs#L15
- `interface IComIID` with `static` property that returns the Guid: https://github.com/microsoft/CsWin32/blob/990484777aa161aebf62ee02ba457b2f5f8286cc/src/Microsoft.Windows.CsWin32/Generator.Com.cs#L1039-L1042

**Comment by AaronRobinsonMSFT** [comment]
> The same can be said about number of other COM related constants that one needs for most real-world COM scenarios. (For example, the E_... error codes.)

That is fair about the entire `HRESULT` surface area, but in .NET we have exceptions that encapsulate those ideas (for example,  [`Exception.HResult`](https://learn.microsoft.com/dotnet/api/system.exception.hresult)). There is nothing in .NET that attempts to encapsulate the bare minimum of a COM server. This means base class identifiers like `IUnknown`, `IClassFactory` and `IClassFactory2` are helpful. An argument could be made the class factory IIDs are unnecessary and that does make sense. The `IUnknown` though comes up and will continue to come up as we lean more in COM source generation.

I see an argument for this being the only IID we provide.

**Comment by jkotas** [comment]
I do not see a problem with the COM interop source generator embedding own private copy of the IUnknown Guid in the shape that it finds convenient. The COM interop source generators need to embed many Guids and other constants, saving it from embedding one less Guid does not move the needle.

**Comment by AaronRobinsonMSFT** [comment]
> saving it from embedding one less Guid does not move the needle.

I don't think this is about moving a needle or trying to address some monumental short coming. The point of this was merely to handle the case because almost every COM interop library makes this define, it is systemic almost across all libraries, see terrafx or WinForms, when doing something outside of or low-level with the built-in COM system. Providing the most fundamental identifier in the entire COM system would appear to be a simple yet convenient constant for those who want to build one up. Even on the `ComWrappers` API we provide the functions to implement `IUnknown`, the option to override `IUnknown`, yet not how to say `IUnknown` in the system.

**Comment by jkotas** [comment]
CsWin32 (used by WinForms) and Terrafx have `IID_IUnknown` naturally auto-generated just like any other IID. I do not expect that the BCL definition would be of much use for them.

I can see that the BCL definition can be useful for manual bespoke IUnknown implementations, like the one in https://github.com/dotnet/performance/pull/3298#discussion_r1306271780 that motivated this issue. If the convenience of bespoke IUnknown implementations is the scenario, we may also want to expose other constants that are always needed for that: `S_OK`, `E_NOINTERFACE ` and layout of IUnknown vtable.

**Comment by AaronRobinsonMSFT** [comment]
I don't think the vtable is all that helpful and goes against what the runtime has been modeling with respect to vtable creation by using `IntPtr*` or `void**`. The detail provided is just going to encourage large architecture schemes for defining a slew of vtable types that have little practical value other than a desire to architect systems.

**Comment by tannergooding** [comment]
> due to alignment concerns

Not alignment, but rather endianess.

> Do we have measurements that show this has a performance impact?

For TerraFX it was impactful namely because I had thousands of IIDs (currently 7112, at least). This generated an absolutely massive static constructor that was just initializing constant data.

So switching to `static ref readonly` properties massively reduced startup cost and saved overall assembly space, at the cost of baking in an implementation detail (which was fine since Windows is little-endian only today).

It had an indirect benefit in that I could rely on the data being guaranteed pinned and therefore `Unsafe.AsPointer` being safe. But that's also not quite as important.

**Comment by tannergooding** [comment]
> I do not expect that the BCL definition would be of much use for them.

TerraFX would indeed continue generating its own copy. For both consistency and convenience around some of the indirect benefits I mentioned.

**Comment by Sergio0694** [comment]
> "Not alignment, but rather endianess."

Ah, right. Is this something we can now work around with the new support for RVA spans?
Or would this require the full blown "constant data spans" thing that's not implemented yet? 🤔

**Comment by tannergooding** [comment]
It'd require the new feature, since you have a struct with fields of multiple different types (`uint, ushort, ushort, byte[8]`) and so you can't represent that with a single endianness fixup. You need something that does fixups per field.
